cmd/metal-api/internal/scaler/poolscaler.go

```diff
@@ -0,0 +1,178 @@
+package scaler
+
+import (
+	"math"
+	"math/rand"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/metal-stack/metal-api/cmd/metal-api/internal/metal"
+	"go.uber.org/zap"
+)
+
+type (
+	PoolScaler struct {
+		log     *zap.SugaredLogger
+		manager MachineManager
+	}
+	PoolScalerConfig struct {
+		Log     *zap.SugaredLogger
+		Manager MachineManager
+	}
+	MachineManager interface {
+		AllMachines() (metal.Machines, error)
+		WaitingMachines() (metal.Machines, error)
+		ShutdownMachines() (metal.Machines, error)
+
+		Shutdown(m *metal.Machine) error
+		PowerOn(m *metal.Machine) error
+	}
+)
+
+func NewPoolScaler(log *zap.SugaredLogger, manager MachineManager) *PoolScaler {
+	return &PoolScaler{
+		log:     log,
```

Comment:
If the pool scaler gets the partition for the constructor, you can even add this to all logs by initializing the log adding `.With()`.